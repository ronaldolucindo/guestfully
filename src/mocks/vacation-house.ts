import houseImage from '@/assets/vacation-house.jpg'
import bedroom from '@/assets/bedroom.jpg'
import livingroom from '@/assets/living-room.jpg'
import kitchen from '@/assets/kitchen.jpg'

const VACATION_HOUSE_MOCK = {
  name: 'Mar Azul House',
  subtitle: 'Beach house with an amazing view to the ocean',
  address: '1010 Main Street, Salvador - Bahia - Brazil',
  guests: 6,
  bedrooms: 3,
  bathrooms: 3,
  price: 989.99,
  amenities: [
    'toaster',
    'toiletries',
    'internet',
    'towels',
    'tv',
    'surveillance',
    'washer/dryer',
    'blender',
    'conditioner',
    'desk',
    'fridge',
    'wardrobe',
  ],
  description:
    'Welcome to Mar Azul House, your private oasis by the turquoise waters of Salvadorâ€™s coastline. Nestled amidst swaying palm trees and overlooking the pristine Barra Beach, this beachfront retreat offers an idyllic escape for travelers seeking tranquility and sun-kissed days.',
  images: [
    {
      original: houseImage,
      thumbnail: houseImage,
    },
    {
      original: bedroom,
      thumbnail: bedroom,
    },
    {
      original: livingroom,
      thumbnail: livingroom,
    },
    {
      original: kitchen,
      thumbnail: kitchen,
    },
  ],
}
export default VACATION_HOUSE_MOCK
