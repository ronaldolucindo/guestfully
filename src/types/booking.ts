import { Property } from './property'

export type Booking = {
  id: string
  guests: number
  note: string
  totalPrice: number
  date: [Date, Date]
  property: {
    id: Property['id']
    name: Property['name']
    address: Property['address']
    image: string
  }
}

export type DateInterval = { start: Date; end: Date }
