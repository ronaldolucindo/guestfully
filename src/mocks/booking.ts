import { Booking } from '@/types/booking'

export const BOOKING_MOCK: Booking = {
  id: 'id',
  guests: 3,
  note: 'my notes',
  totalPrice: 900.9877,
  date: [new Date(), new Date()],
  property: {
    id: 'pId',
    name: 'pName',
    address: 'pAddress',
    image: 'imageUrl',
  },
}
