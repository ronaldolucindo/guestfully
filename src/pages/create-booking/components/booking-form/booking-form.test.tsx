/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { render, screen, waitFor } from '@testing-library/react'
import BookingForm from './booking-form'
import { vitest } from 'vitest'
import { BOOKING_MOCK } from '@/mocks/booking'
import VACATION_HOUSE_MOCK from '@/mocks/vacation-house'
import userEvent from '@testing-library/user-event'

const formProps = {
  booking: BOOKING_MOCK,
  property: VACATION_HOUSE_MOCK,
  bookedDates: [{ start: new Date(), end: new Date() }],
  onSubmit: vitest.fn(),
}
const FORMATTED_PRICE = '$989.99'

describe('BookingForm', () => {
  it('displays input initial value', () => {
    render(<BookingForm {...formProps} />)

    expect(screen.getByLabelText('Guests')).toHaveValue(
      `${formProps.booking.guests}`,
    )
    expect(screen.getByLabelText('Note')).toHaveValue(formProps.booking.note)
  })
  it('does not display input initial value', () => {
    render(
      <BookingForm
        property={formProps.property}
        onSubmit={formProps.onSubmit}
        bookedDates={formProps.bookedDates}
      />,
    )

    expect(screen.getByLabelText('Note')).not.toHaveValue()
  })
  it('it shows price per night', () => {
    render(<BookingForm {...formProps} />)

    expect(screen.getByText(`${FORMATTED_PRICE}`)).toBeTruthy()
  })
  it('triggers submit function', async () => {
    const user = userEvent.setup()
    render(<BookingForm {...formProps} />)

    await waitFor(() => {
      const saveBtn = screen.getByTestId('BookingForm.bookBtn')
      expect(saveBtn).toBeInTheDocument()

      user.click(saveBtn)
      expect(formProps.onSubmit).toHaveBeenCalled()
    })

    expect(screen.getByLabelText('Guests')).toHaveValue(
      `${formProps.booking.guests}`,
    )
    expect(screen.getByLabelText('Note')).toHaveValue(formProps.booking.note)
  })
})
