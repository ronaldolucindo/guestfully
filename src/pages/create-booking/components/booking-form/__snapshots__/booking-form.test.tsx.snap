// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BookingForm > matches snapshot 1`] = `
<form
  class="w-full"
  data-testid="BookingForm"
  novalidate=""
>
  <p
    class="text-lg font-bold"
  >
    $989.99
    <span
      class="text-sm font-bold"
    >
      / night
    </span>
  </p>
  <div />
  <div
    class="my-2"
    data-headlessui-state=""
  >
    <p
      class="text-sm font-bold"
    >
      Date*
    </p>
    <div
      class="w-full"
    >
      <div
        class="react-datepicker-wrapper w-full"
      >
        <div
          class="react-datepicker__input-container"
        >
          <input
            class="mt-1 block w-full rounded-lg bg-slate-100 px-3 py-2 text-sm/6 data-[focus]:outline-2 data-[focus]:outline-white/25 "
            name="date"
            type="text"
            value="07/18/2024 - 07/18/2024"
          />
          <button
            aria-label="Close"
            class="react-datepicker__close-icon"
            tabindex="-1"
            type="button"
          />
        </div>
      </div>
    </div>
    <span
      hidden=""
      style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
    />
  </div>
  <div
    data-headlessui-state=""
  >
    <label
      class="text-sm font-bold"
      data-headlessui-state=""
      for="headlessui-control-:r12:"
      id="headlessui-label-:r13:"
    >
      Guests
    </label>
    <div
      class="relative"
    >
      <select
        aria-labelledby="headlessui-label-:r13:"
        class="mt-3 block w-full appearance-none rounded-lg bg-slate-100 px-3 py-2 focus:outline-none data-[focus]:outline-2 data-[focus]:outline-white/25"
        data-headlessui-state=""
        id="headlessui-control-:r12:"
        name="guests"
      >
        <option
          value="1"
        >
          1
        </option>
        <option
          value="2"
        >
          2
        </option>
        <option
          value="3"
        >
          3
        </option>
        <option
          value="4"
        >
          4
        </option>
        <option
          value="5"
        >
          5
        </option>
        <option
          value="6"
        >
          6
        </option>
      </select>
      <svg
        aria-hidden="true"
        class="group pointer-events-none absolute right-2.5 top-2.5 size-4"
        data-slot="icon"
        fill="currentColor"
        viewBox="0 0 16 16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M4.22 6.22a.75.75 0 0 1 1.06 0L8 8.94l2.72-2.72a.75.75 0 1 1 1.06 1.06l-3.25 3.25a.75.75 0 0 1-1.06 0L4.22 7.28a.75.75 0 0 1 0-1.06Z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <span
      hidden=""
      style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
    />
  </div>
  <div
    class="my-2"
    data-headlessui-state=""
  >
    <label
      class="text-sm font-bold"
      data-headlessui-state=""
      for="headlessui-control-:r15:"
      id="headlessui-label-:r16:"
    >
      Note
    </label>
    <textarea
      aria-labelledby="headlessui-label-:r16:"
      class="mt-1 block w-full rounded-lg bg-slate-100 px-3 py-2 text-sm/6 data-[focus]:outline-2 data-[focus]:outline-white/25"
      data-headlessui-state=""
      id="headlessui-control-:r15:"
    >
      my notes
    </textarea>
    <span
      hidden=""
      style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
    />
  </div>
  <button
    class="rounded-full w-full bg-blue-700 px-4 py-2 text-white data-[active]:bg-blue-700 data-[disabled]:bg-gray-500 data-[hover]:bg-blue-800"
    data-headlessui-state=""
    data-testid="BookingForm.bookBtn"
    type="submit"
  >
    Book now
  </button>
</form>
`;
